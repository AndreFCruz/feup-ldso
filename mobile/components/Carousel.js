import React from 'react';
import { Animated, View, ScrollView, StyleSheet, Dimensions, TouchableOpacity} from 'react-native';
import {Icon} from 'native-base';
import PropTypes from 'prop-types';
import { widthPercentageToDP as wp, heightPercentageToDP as hp } from 'react-native-responsive-screen';
import facultyStyles from '../constants/SpecificStyles';

const { width } = Dimensions.get('window');
const FIXED_BAR_WIDTH = 280;
const BAR_SPACE = 10;


class Carousel extends React.Component {

    numItems = this.props.videos.length
    itemWidth = (FIXED_BAR_WIDTH / this.numItems) - ((this.numItems - 1) * BAR_SPACE)
    animVal = new Animated.Value(0)


    scrollToPreviousPage = () => {
        this.scroller.scrollTo({x: -width, y: 0});
    };

    scrollToNextPage = () => {
        this.scroller.scrollTo({x: width, y: 0});
    };

    trackDynamicStyle = function(itemWidth, i) {
        return {
            width: itemWidth,
            marginLeft: i === 0 ? 0 : BAR_SPACE
        };
    }

    animatedViewDynamicStyle = function(itemWidth, scrollBarVal) {
        return{
            width: itemWidth,
            transform: [
                { translateX: scrollBarVal },
            ],
        };
    }

    render() {
        const { videos, name } = this.props;

        let barArray = [];
        for(let i=0; i<this.numItems; i++) {
            const scrollBarVal = this.animVal.interpolate({
                inputRange: [width * (i - 1), width * (i + 1)],
                outputRange: [-this.itemWidth, this.itemWidth],
                extrapolate: 'clamp',
            });
            
            const thisBar = (
                <View
                    key={`bar${i}`}
                    style={[ styles.track,this.trackDynamicStyle(this.itemWidth, i)]}
                >
                    <Animated.View
                        style={[styles.bar, facultyStyles[name].carouselBar,this.animatedViewDynamicStyle(this.itemWidth, scrollBarVal)]}
                    />
                </View>
            );
            barArray.push(thisBar);
        }

        if (videos && videos.length) {
            return (
                <View>
                    <View style={styles.scrollContainer}>
                        <ScrollView
                            horizontal
                            pagingEnabled
                            showsHorizontalScrollIndicator={false}
                            scrollEventThrottle={10}
                            onScroll={
                                Animated.event(
                                    [{ nativeEvent: { contentOffset: { x: this.animVal } } }]
                                )
                            }
                            ref={(scroller) => {this.scroller = scroller;}}
                        >
                            {videos}
                        </ScrollView>
                        <View style={styles.barContainer}>
                            {barArray}
                        </View>
                    </View>
                    <View style={[styles.leftArrow]}>
                        <TouchableOpacity onPress={this.scrollToPreviousPage}>
                            <Icon style={[styles.leftArrowIcon, facultyStyles[name].videoArrow]} type="FontAwesome" name="angle-left" />
                        </TouchableOpacity>
                    </View>
                    <View style={styles.rightArrow}>
                        <TouchableOpacity onPress={this.scrollToNextPage}>
                            <Icon style={[styles.rightArrowIcon, facultyStyles[name].videoArrow]} type="FontAwesome" name="angle-right" />
                        </TouchableOpacity>
                    </View>
                </View>
            );
        }
        return null;
    }
}

Carousel.propTypes = {
    name: PropTypes.string,
    videos: PropTypes.array
};

const styles = StyleSheet.create({
    scrollContainer: {
        flex: 1,
        height: hp('100%'),
        justifyContent: 'center',
        alignItems: 'center'
    },
    barContainer: {
        zIndex: 2,
        position: 'absolute',
        top: hp('87.5%'),
        justifyContent: 'center',
        flexDirection: 'row'
    },
    track: {
        backgroundColor: '#d3d3d3',
        overflow: 'hidden',
        height: 2
    },
    bar: {
        height: 2,
        position: 'absolute',
        left: 0,
        top: 0,
    },
    leftArrow: {
        zIndex: 2,
        position: 'absolute',
        top: hp('40%'),
        left: wp('1%')
    },
    leftArrowIcon: {
        fontSize: wp('7%')
    },
    rightArrow: {
        zIndex: 2,
        position: 'absolute',
        top: hp('40%'),
        left: wp('97%'),
    },
    rightArrowIcon: {
        fontSize: wp('7%')
    },


});

export default Carousel;