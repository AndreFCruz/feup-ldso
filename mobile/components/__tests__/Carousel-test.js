
import {View} from 'react-native';
import React from 'react';
import Carousel  from '../Carousel';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });


function setup() {
    const views = [<View key={0}/>, <View key={1}/>];
    const wrapper = shallow(<Carousel videos={views} name='feup'/>);
    return wrapper;

}
describe('Rendering ', () => {
    it('renders carousel with content received', () => {
        const wrapper = setup();
        expect(wrapper.dive()).toMatchSnapshot();
    });
    it('renders carousel without content', () => {
        const views = [];
        const wrapper = shallow(<Carousel videos={views} name='feup'/>,
            {
                disableLifecycleMethods: true
            });
        expect(wrapper.type()).toEqual(null);
    });

});