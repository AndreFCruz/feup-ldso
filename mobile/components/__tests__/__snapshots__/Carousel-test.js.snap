// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering  renders carousel with content received 1`] = `
<View>
  <Component
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "height": 1334,
        "justifyContent": "center",
      }
    }
  >
    <ScrollViewMock
      horizontal={true}
      onScroll={[Function]}
      pagingEnabled={true}
      scrollEventThrottle={10}
      showsHorizontalScrollIndicator={false}
    >
      <Component
        key="0"
      />
      <Component
        key="1"
      />
    </ScrollViewMock>
    <Component
      style={
        Object {
          "flexDirection": "row",
          "justifyContent": "center",
          "position": "absolute",
          "top": 1167.5,
          "zIndex": 2,
        }
      }
    >
      <Component
        key="bar0"
        style={
          Array [
            Object {
              "backgroundColor": "#d3d3d3",
              "height": 2,
              "overflow": "hidden",
            },
            Object {
              "marginLeft": 0,
              "width": 130,
            },
          ]
        }
      >
        <AnimatedComponent
          style={
            Array [
              Object {
                "height": 2,
                "left": 0,
                "position": "absolute",
                "top": 0,
              },
              Object {
                "backgroundColor": "#9f2d20",
              },
              Object {
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                ],
                "width": 130,
              },
            ]
          }
        />
      </Component>
      <Component
        key="bar1"
        style={
          Array [
            Object {
              "backgroundColor": "#d3d3d3",
              "height": 2,
              "overflow": "hidden",
            },
            Object {
              "marginLeft": 10,
              "width": 130,
            },
          ]
        }
      >
        <AnimatedComponent
          style={
            Array [
              Object {
                "height": 2,
                "left": 0,
                "position": "absolute",
                "top": 0,
              },
              Object {
                "backgroundColor": "#9f2d20",
              },
              Object {
                "transform": Array [
                  Object {
                    "translateX": -130,
                  },
                ],
                "width": 130,
              },
            ]
          }
        />
      </Component>
    </Component>
  </Component>
  <Component
    style={
      Array [
        Object {
          "left": 7.5,
          "position": "absolute",
          "top": 533.5,
          "zIndex": 2,
        },
      ]
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Styled(Icon)
        name="angle-left"
        style={
          Array [
            Object {
              "fontSize": 52.5,
            },
            Object {
              "color": "#9f2d20",
            },
          ]
        }
        type="FontAwesome"
      />
    </TouchableOpacity>
  </Component>
  <Component
    style={
      Object {
        "left": 727.5,
        "position": "absolute",
        "top": 533.5,
        "zIndex": 2,
      }
    }
  >
    <TouchableOpacity
      activeOpacity={0.2}
      onPress={[Function]}
    >
      <Styled(Icon)
        name="angle-right"
        style={
          Array [
            Object {
              "fontSize": 52.5,
            },
            Object {
              "color": "#9f2d20",
            },
          ]
        }
        type="FontAwesome"
      />
    </TouchableOpacity>
  </Component>
</View>
`;
